src/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs

```diff
@@ -0,0 +1,866 @@
+        }
+
+        internal void OnReaderCompleted(Action<Exception, object> callback, object state)
+        {
+            if (callback == null)
+            {
+                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.callback);
+            }
+
+            PipeCompletionCallbacks completionCallbacks;
+            lock (_sync)
+            {
+                completionCallbacks = _readerCompletion.AddCallback(callback, state);
+            }
+
+            if (completionCallbacks != null)
+            {
+                TrySchedule(_writerScheduler, _invokeCompletionCallbacks, completionCallbacks);
+            }
+        }
+
+        internal PipeAwaiter<ReadResult> ReadAsync(CancellationToken token)
+        {
+            CancellationTokenRegistration cancellationTokenRegistration;
+            if (_readerCompletion.IsCompleted)
+            {
+                ThrowHelper.ThrowInvalidOperationException_NoReadingAllowed();
+            }
+            lock (_sync)
+            {
+                cancellationTokenRegistration = _readerAwaitable.AttachToken(token, _signalReaderAwaitable, this);
+            }
+            cancellationTokenRegistration.Dispose();
+            return new PipeAwaiter<ReadResult>(_reader);
+        }
+
+        internal bool TryRead(out ReadResult result)
+        {
+            lock (_sync)
+            {
+                if (_readerCompletion.IsCompleted)
+                {
+                    ThrowHelper.ThrowInvalidOperationException_NoReadingAllowed();
+                }
+
+                result = new ReadResult();
+                if (_length > 0 || _readerAwaitable.IsCompleted)
+                {
+                    GetResult(ref result);
+                    return true;
+                }
+
+                if (_readerAwaitable.HasContinuation)
+                {
+                    ThrowHelper.ThrowInvalidOperationException_AlreadyReading();
+                }
+                return false;
+            }
+        }
+
+        private static void TrySchedule(PipeScheduler scheduler, Action action)
+        {
+            if (action != null)
+            {
+                scheduler.Schedule(action);
+            }
+        }
+
+        private static void TrySchedule(PipeScheduler scheduler, Action<object> action, object state)
+        {
+            if (action != null)
+            {
+                scheduler.Schedule(action, state);
+            }
+        }
+
+        private void CompletePipe()
+        {
+            lock (_sync)
+            {
+                if (_disposed)
+                {
+                    return;
+                }
+
+                _disposed = true;
+                // Return all segments
+                // if _readHead is null we need to try return _commitHead
+                // because there might be a block allocated for writing
+                BufferSegment segment = _readHead ?? _commitHead;
+                while (segment != null)
```

Comment:
Can this be re-written without creating a local copy of BufferSegment at each iteration?
```C#
while (segment.NextSegment != null)
{
    segment.ResetMemory();
    segment = segment.NextSegment;
}
```

Reply:
`    segment.ResetMemory();` resets `NextSegment`